pkg/clusteragent/admission/mutate/cwsinstrumentation/k8scp/cp.go

```diff
@@ -0,0 +1,169 @@
+}
+
+// NewCopy creates a Command instance
+func NewCopy(config *restclient.Config, clientSet kubernetes.Interface) *Copy {
+	config.APIPath = "/api"
+	config.GroupVersion = &schema.GroupVersion{Version: "v1"}
+	config.NegotiatedSerializer = serializer.WithoutConversionCodecFactory{CodecFactory: scheme.Codecs}
+	ioStreams, in, out, errOut := genericiooptions.NewTestIOStreams()
+	return &Copy{
+		streams:   ioStreams,
+		in:        in,
+		out:       out,
+		errOut:    errOut,
+		config:    config,
+		clientSet: clientSet,
+	}
+}
+
+func (o *Copy) CopyToPod(localFile string, remoteFile string, pod *corev1.Pod, container string) error {
+	// sanity check
+	if _, err := os.Stat(localFile); err != nil {
+		return fmt.Errorf("%s doesn't exist in local filesystem", localFile)
+	}
+
+	reader, writer := io.Pipe()
+	srcFile := newLocalPath(localFile)
+	destFile := newRemotePath(remoteFile)
+
+	go func(src localPath, dest remotePath, writer io.WriteCloser) {
+		defer writer.Close()
+		if err := makeTar(src, dest, writer); err != nil {
+			log.Debugf("failed to tar local file: %v", err)
+		}
+	}(srcFile, destFile, writer)
+
+	// arguments are split on purpose to differentiate this kubectl cp from others
+	cmdArr := []string{"tar", "-x", "-m", "-f", "-"}
+	destFileDir := destFile.Dir().String()
+	if len(destFileDir) > 0 {
+		cmdArr = append(cmdArr, "-C", destFileDir)
+	}
+
+	options := &ExecOptions{}
+	options.StreamOptions = StreamOptions{
+		IOStreams: genericiooptions.IOStreams{
+			In:     reader,
+			Out:    o.out,
+			ErrOut: o.errOut,
+		},
+		Stdin:         true,
+		Namespace:     pod.GetNamespace(),
+		PodName:       pod.GetName(),
+		ContainerName: container,
+	}
+	options.Command = cmdArr
+	options.Config = o.config
+	options.Executor = &DefaultRemoteExecutor{}
+
+	return options.Run(pod)
+}
+
+func makeTar(src localPath, dest remotePath, writer io.Writer) error {
+	tarWriter := tar.NewWriter(writer)
+	defer tarWriter.Close()
+
+	srcPath := src.Clean()
+	destPath := dest.Clean()
+	return recursiveTar(srcPath.Dir(), srcPath.Base(), destPath.Dir(), destPath.Base(), tarWriter)
+}
+
+func recursiveTar(srcDir, srcFile localPath, destDir, destFile remotePath, tw *tar.Writer) error {
+	matchedPaths, err := srcDir.Join(srcFile).Glob()
+	if err != nil {
+		return err
+	}
+	for _, fpath := range matchedPaths {
+		stat, err := os.Lstat(fpath)
+		if err != nil {
+			return err
+		}
+		if stat.IsDir() {
+			files, err := os.ReadDir(fpath)
+			if err != nil {
+				return err
+			}
+			if len(files) == 0 {
+				//case empty directory
+				hdr, _ := tar.FileInfoHeader(stat, fpath)
+				hdr.Name = destFile.String()
+				if err := tw.WriteHeader(hdr); err != nil {
+					return err
+				}
+			}
+			for _, f := range files {
+				if err := recursiveTar(srcDir, srcFile.Join(newLocalPath(f.Name())),
+					destDir, destFile.Join(newRemotePath(f.Name())), tw); err != nil {
+					return err
+				}
+			}
+			return nil
+		} else if stat.Mode()&os.ModeSymlink != 0 {
```

Comment:
nit: i like using the indent flow [here](https://go.dev/wiki/CodeReviewComments#indent-error-flow), that'd remove the else and many braces, eg:

```go
}
if stat.Mode()&os.ModeSymlink != 0 {
```

Reply:
Done 👍🏻